mlx/backend/metal/allocator.cpp

```diff
@@ -29,6 +29,7 @@ BufferCache::BufferCache(MTL::Device* device)
     : device_(device), head_(nullptr), tail_(nullptr), pool_size_(0) {}
 
 BufferCache::~BufferCache() {
+  auto thread_pool = metal::new_scoped_memory_pool();
```

Comment:
Top level calls that involve autoreleased objects (e.g. any time it interacts with metal-cpp) require an autoreleasepool to be in place.  This replaces the `global_memory_pool` and adds the benefit that it actually cleans up the objects on return rather than abandoning them (leak).